lib/types/src/features.rs

```diff
@@ -264,6 +265,126 @@ impl Features {
             && (!required.relaxed_simd || self.relaxed_simd)
             && (!required.extended_const || self.extended_const)
     }
+
+    /// Detects required WebAssembly features from a module binary.
+    ///
+    /// This method analyzes a WebAssembly module's binary to determine which
+    /// features it requires. It does this by:
+    /// 1. Attempting to validate the module with different feature sets
+    /// 2. Analyzing validation errors to detect required features
+    /// 3. Parsing the module to detect certain common patterns
+    ///
+    /// # Arguments
+    ///
+    /// * `wasm_bytes` - The binary content of the WebAssembly module
+    ///
+    /// # Returns
+    ///
+    /// A new `Features` instance with the detected features enabled.
+    pub fn detect_from_wasm(wasm_bytes: &[u8]) -> Result<Self, wasmparser::BinaryReaderError> {
```

Comment:
This needs to be either in wasmer or wasmer-cli to not force the wasmparser dependency on wasmer_types.

Once we move it to wasmer, we should put it behind a feature flag (as wasmparser might not always be available), which should be named `detect-wasm-features`

Reply:
It's a utility similar to `wat`/`wat2wasm`

Reply:
TL;DR this is needed by package, so we can't move to wasmer. We will feature flag this function for now

Reply:
@xdoardo did the feature flag, resolving and opening a linear ticket for it.